### PR TITLE
feat(streaming): add EventStreamer for iterating stream events

### PR DESCRIPTION
**Pull request by ajbozarth**
## Issue

Follow-up to the design discussion on #1543 (now merged) — streaming event consumption ergonomics.

## Description
Adds an opt-in `EventStreamer`, returned by `stream(..., as_events=True)`, for iterating a single stream's typed `StreamEvent` objects directly with `async for`. The only other route to those events is the process-global `STREAMING_EVENT` hook, which is awkward for a single stream — you register a plugin and demultiplex on `streaming_id`.

Under the hood it wraps a `Streamer`, drives it to completion on a background task, and delivers the events through a queue the consumer drains with `async for`.

It is additive and opt-in: `stream()`'s default chunk-iterating `Streamer` is unchanged, and the `STREAMING_EVENT` hook still fires on every stream, so telemetry and multi-stream observers are unaffected. `EventStreamer` is just the ergonomic way to consume one stream's events inline; the hook remains the right tool for observing many at once.

The streaming examples and docs go back to iterating events, reverting the `streaming_event`-hook workaround #1543 adopted when the iterator was removed; the v0.7→v0.8 migration guide now maps `result.events()` to `stream(as_events=True)` to match.

### Testing
- [x] Tests added to the respective file if code was changed
- [ ] New code has 100% coverage if code was added
- [ ] Ensure existing tests and github automation passes (a maintainer will kick off the github automation when the rest of the PR is populated)

Unit tests cover `EventStreamer`'s event-mode paths (natural completion, requirement failure, error re-raise, early break), resume-and-drain after a break, teardown edge cases (idempotent `aclose`, a faulted/abandoned pump, a `BaseException` during setup that must not hang), and setup-error surfacing; `test/typing/check_stream_as_events.py` covers the `as_events` overloads including a dynamic `bool`.

### Attribution
- [x] AI coding assistants used

---

### Adding a new component, requirement, sampling strategy, or tool?

If your PR adds or modifies one of the types below, check the matching box. A checklist of type-specific review items will be posted as a comment.

- [ ] Component
- [ ] Requirement
- [ ] Sampling Strategy
- [ ] Tool

NOTE: Please ensure you have an issue that has been acknowledged by a core contributor and routed you to open a pull request against this repository. Otherwise, please open an issue before continuing with this pull request.

### CODE REVIEW

**Comment by ajbozarth** [comment]
@jakelorocco here's my initial draft POC for a events iterator for streaming. This is the better of the two POCs I wrote and I'll leave this here for your review while I'm out. Once I get back next week we can sync and I'll continue work on this

File: mellea/stdlib/streaming.py [hunk 1]
```diff
@@ -330,15 +335,147 @@ async def __aexit__(self, *exc_info: object) -> None:
...
+                ctx,
+                chunking=chunking,
+                requirements=requirements,
+                validation_backend=validation_backend,
+                event_queue=self._queue,
+            )
+            async for _ in self._streamer:  # chunks discarded; events are surfaced
+                pass
+        finally:
+            self._queue.put_nowait(None)
```
Comment (jakelorocco) [inline mellea/stdlib/streaming.py:399]: Since we are awaiting the underlying _stream function, the stream still progresses even when not directly awaiting on this event streamer, right?

File: mellea/stdlib/streaming.py [hunk 2]
```diff
@@ -330,15 +335,147 @@ async def __aexit__(self, *exc_info: object) -> None:
...
+        try:
+            self._streamer = await _stream(
+                action,
+                backend,
+                ctx,
+                chunking=chunking,
+                requirements=requirements,
+                validation_backend=validation_backend,
+                event_queue=self._queue,
+            )
```
Comment (jakelorocco) [inline mellea/stdlib/streaming.py:384]: Is it not an option to do the as_events boolean on the streamer itself and remove the need for discarding empty chunks and creating a new queue?

**Review by jakelorocco — COMMENTED** [review]
This pretty much looks like exactly what I was looking for; thank you!

File: mellea/stdlib/streaming.py (hunk 1 above)
Comment (ajbozarth) [reply mellea/stdlib/streaming.py:399]: That is correct, just like how the old event queue worked

File: mellea/stdlib/streaming.py (hunk 2 above)
Comment (ajbozarth) [reply mellea/stdlib/streaming.py:384]: This was actually the first implementation I pursued, but it ran into issues handling the error paths since they happen after the iterator is "done" so the iterator can't emit them. That's when I pivoted to this implementation

**Comment by ajbozarth** [comment]
PR is updated and moved out of draft POC, ready for review

File: mellea/stdlib/streaming.py [hunk 3]
```diff
@@ -330,15 +337,179 @@ async def __aexit__(self, *exc_info: object) -> None:
...
+        except BaseException as exc:
+            # Resolve _ready on every setup exit so `await _ready` can't hang.
+            if isinstance(exc, Exception):
+                self._ready.set_exception(exc)
+            else:
+                self._ready.cancel()
```
Comment (planetf1) [inline mellea/stdlib/streaming.py:388]: Could this preserve non-cancellation `BaseException`s rather than cancelling `_ready`? If `_stream()` raises a custom `BaseException`, `KeyboardInterrupt`, or `SystemExit` during setup, awaiting `stream(..., as_events=True)` receives `CancelledError` instead. The caller cannot see or handle the real failure, even though the default `stream()` path preserves it; `stream()` then retrieves the original pump exception only to discard it. I suggest considering reserving `_ready.cancel()` for `asyncio.CancelledError`, using `_ready.set_exception(exc)` for other exceptions, and updating the regression test to expect the original exception.

File: mellea/stdlib/streaming.py [hunk 4]
```diff
@@ -330,15 +337,179 @@ async def __aexit__(self, *exc_info: object) -> None:
...
+    `async with`. It is its own iterator, so `break` then iterating again
+    resumes from the next queued event. Events are produced eagerly, independent
+    of consumption, so the stream can run ahead of iteration; outcome attributes
+    (`full_text`, `completed_normally`, …) mirror the wrapped `Streamer`. Do not
+    instantiate directly.
+    """
+
+    def __init__(self) -> None:
+        """Create an idle handle; `stream()` spawns the pump task."""
+        self._queue: asyncio.Queue[StreamEvent | None] = asyncio.Queue()
```
Comment (planetf1) [inline mellea/stdlib/streaming.py:358]: This queue is unbounded while the pump drains the backend independently of the consumer. A slow or stalled consumer retains every event in memory; with `chunking=None`, that includes one `ChunkEvent` per raw delta. A sufficiently long response can therefore grow the process's memory until it is killed, rather than applying backpressure to generation. Could we define a bounded, cancellation-safe buffering policy here? A simple `maxsize` needs a close-path test too, so terminal events and teardown cannot block when the consumer has stopped.

File: docs/docs/how-to/use-async-and-streaming.md [hunk 5]
```diff
@@ -310,9 +310,36 @@ async def main() -> None:
...
+instead, pass `as_events=True` to `stream()`: it returns an `EventStreamer` you
+iterate the same way, each step yielding an event rather than a chunk:
+
+```python
+async with await stream(
+    action, m.backend, m.ctx, requirements=[req], chunking="sentence", as_events=True
+) as streamer:
+    async for event in streamer:
+        match event:
+            case ChunkEvent():
```
Comment (planetf1) [inline docs/docs/how-to/use-async-and-streaming.md:333]: The preceding example imports only `stream`, so a reader who follows it gets `NameError` on the first `ChunkEvent` (and likewise for the other event classes). Could you consider importing the event types in the preceding import block or directly before this example?

**Review by planetf1 — CHANGES_REQUESTED** [review]
I found three correctness and documentation issues inline.

File: mellea/stdlib/streaming.py (hunk 3 above)
Comment (ajbozarth) [reply mellea/stdlib/streaming.py:388]: fixed in https://github.com/generative-computing/mellea/pull/1567/commits/6a65309edd97513934c1f926b4f1e4278de3b75a

File: docs/docs/how-to/use-async-and-streaming.md (hunk 5 above)
Comment (ajbozarth) [reply docs/docs/how-to/use-async-and-streaming.md:333]: fixed in https://github.com/generative-computing/mellea/pull/1567/commits/6a65309edd97513934c1f926b4f1e4278de3b75a

File: mellea/stdlib/streaming.py (hunk 4 above)
Comment (ajbozarth) [reply mellea/stdlib/streaming.py:358]: Thanks — I dug into this and I don't think it's an issue in practice.

**Per-stream cost is small and bounded.** The pump drives the stream to completion and stops, so the buffer ceiling is one response's worth of events — O(output length), itself capped by the model's max output. Measured, fully stalled consumer, `chunking=None`:

| output | heap growth |
|---|---|
| 1k tokens | ~0.18 MB |
| 10k tokens | ~1.6 MB |
| 50k tokens | ~7.7 MB |

~160 bytes/event, linear — OOM'ing a 2 GB process from one stream would take ~13M output tokens, well past any model. This is also the same eager, unbounded queue the phase-1 `events()` iterator used (#1095), so it's an established pattern here, not a regression.

**A bounded queue would be worse here, not better.** The only way `maxsize` applies real backpressure is a blocking `put`, which stops the pump pulling from the backend. For the stalled-consumer case, that converts "bounded memory, generation still completes and the backend stream is drained and released" into "generation blocks indefinitely, the backend connection/model stays pinned, and the stream never completes or populates its outcome attributes." Trading a few MB for an indefinitely-pinned backend generation is the wrong trade for an LLM stream. (Bounded-and-drop instead silently loses events, breaking the iterate-every-event contract.)

So I'd keep the eager, unbounded buffer as-is.

Comment (ajbozarth) [reply mellea/stdlib/streaming.py:358]: I had trouble wording my response so I had Claude write the above with research. My tl;dr is essentially I disagree that we should bound the queue

Comment (planetf1) [reply mellea/stdlib/streaming.py:358]: Thanks Alex — the calcs check out and you've convinced me; unbounded is the right call here. Withdrawing the bounded-queue request.

**Review by planetf1 — APPROVED** [review]
All review threads resolved and verified against 6a65309e:

- BaseException preservation during setup: verified — pump splits on CancelledError vs other exceptions, original error surfaces to the caller, regression test updated to expect it.
- Missing event-type imports in the how-to example: verified.
- Bounded-queue proposal: withdrawn — unbounded eager buffering is the right trade (measured ~160 B/event, O(output length) per stream; a blocking put would pin the backend slot and breaks the close path).

CI green.

**Comment by ajbozarth** [comment]
@jakelorocco if you could give this another review pass before I merge, I want to make sure the edits I made to the POC after your initial approval are still good

File: mellea/stdlib/streaming.py [hunk 6]
```diff
@@ -330,15 +337,178 @@ async def __aexit__(self, *exc_info: object) -> None:
...
+        Prefer `async with stream(..., as_events=True) as s:` so this runs
+        automatically on every exit path.
+        """
+        if self._pump_task is not None:
+            if not self._pump_task.done():
+                self._pump_task.cancel()
+                try:
+                    await self._pump_task
+                except asyncio.CancelledError:
+                    pass  # we cancelled it; consume the resulting CancelledError
```
Comment (jakelorocco) [inline mellea/stdlib/streaming.py]: I think the assumption behind this exception block may be incorrect. It could also be raised if the current task (the one running this aclose) is cancelled from outside of it. In that case, I think we ought to be re-raising the cancelled err.

I don't know how important this is, but I think mots already handle it with their `.cancel_generation` field. Maybe we can do something similar here?

**Review by jakelorocco — COMMENTED** [review]
Found one potential issue; I don't know if it's worth addressing / worth holding up this PR; will defer to you

File: mellea/stdlib/streaming.py (hunk 6 above)
Comment (ajbozarth) [reply mellea/stdlib/streaming.py]: updated to match how MOT handles it and added a test like MOTs as well in https://github.com/generative-computing/mellea/pull/1567/commits/cba531a6f32c9afcd716c267aed4accbc4996ab7
